builder error when unwrapping result of enum's from() (to optional value)

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.4 f603035.c9933da
OS: linux, "EndeavourOS Linux"
Processor: 8 cpus, 64bit, little endian, 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz

getwd: /H/_nobackup/Projects/v_t
vexe: /H/daily/mh/.local/share/v/v
vexe mtime: 2024-02-12 15:34:33

vroot: OK, value: /H/daily/mh/.local/share/v
VMODULES: OK, value: /home/mh/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.43.1
Git vroot status: weekly.2024.06-56-gc9933da6
.git/config present: true

CC version: cc (GCC) 13.2.1 20230801
thirdparty/tcc status: thirdparty-linux-amd64 99683af0

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg main.v`
```v
enum Operation {
	get_area
	get_sector
}

fn main() {
	operation_name := 'get_area'
        // I want 'operation' to have `?Operation' type
	operation := if o := Operation.from(operation_name) { o } else { ?Operation(none) }
}

```
this works:
```v
	operation := if o := Operation.from(operation_name) { ?Operation(o) } else { ?Operation(none) }
```

**What did you expect to see?**

no builder error, probably compiler error instead

**What did you see instead?**
```
/tmp/v_1000/main.01HPEZC36RV5VS0GNSTPAJ2C31.tmp.c:12965: error: cannot convert 'struct _option_main__Operation' to 'enum <anonymous>'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/if.v
+++ b/vlib/v/checker/if.v
@@ -471,13 +471,11 @@ fn (mut c Checker) if_expr(mut node ast.IfExpr) ast.Type {
 							if stmt.typ.is_ptr() != node.typ.is_ptr() {
 								c.error('mismatched types `${c.table.type_to_str(node.typ)}` and `${c.table.type_to_str(stmt.typ)}`',
 									node.pos)
-							} else if stmt.typ.has_flag(.option) || node.typ.has_flag(.option) {
-								if !(stmt.typ.has_flag(.option) && node.typ.has_flag(.option)) {
+							} else if stmt.typ != ast.none_type {
+								if !node.typ.has_flag(.option) && stmt.typ.has_flag(.option) {
 									c.error('mismatched types `${c.table.type_to_str(node.typ)}` and `${c.table.type_to_str(stmt.typ)}`',
 										node.pos)
-								}
-							} else if stmt.typ.has_flag(.result) || node.typ.has_flag(.result) {
-								if !(stmt.typ.has_flag(.result) && node.typ.has_flag(.result)) {
+								} else if !node.typ.has_flag(.result) && stmt.typ.has_flag(.result) {
 									c.error('mismatched types `${c.table.type_to_str(node.typ)}` and `${c.table.type_to_str(stmt.typ)}`',
 										node.pos)
 								}

--- a/vlib/v/checker/if.v
+++ b/vlib/v/checker/if.v
@@ -467,10 +467,21 @@ fn (mut c Checker) if_expr(mut node ast.IfExpr) ast.Type {
 							node.pos)
 					} else {
 						if c.inside_assign && node.is_expr && !node.typ.has_flag(.shared_f)
-							&& stmt.typ.is_ptr() != node.typ.is_ptr()
 							&& stmt.typ != ast.voidptr_type {
-							c.error('mismatched types `${c.table.type_to_str(node.typ)}` and `${c.table.type_to_str(stmt.typ)}`',
-								node.pos)
+							if stmt.typ.is_ptr() != node.typ.is_ptr() {
+								c.error('mismatched types `${c.table.type_to_str(node.typ)}` and `${c.table.type_to_str(stmt.typ)}`',
+									node.pos)
+							} else if stmt.typ.has_flag(.option) || node.typ.has_flag(.option) {
+								if !(stmt.typ.has_flag(.option) && node.typ.has_flag(.option)) {
+									c.error('mismatched types `${c.table.type_to_str(node.typ)}` and `${c.table.type_to_str(stmt.typ)}`',
+										node.pos)
+								}
+							} else if stmt.typ.has_flag(.result) || node.typ.has_flag(.result) {
+								if !(stmt.typ.has_flag(.result) && node.typ.has_flag(.result)) {
+									c.error('mismatched types `${c.table.type_to_str(node.typ)}` and `${c.table.type_to_str(stmt.typ)}`',
+										node.pos)
+								}
+							}
 						}
 					}
 				} else if !node.is_comptime && stmt !in [ast.Return, ast.BranchStmt] {

--- a/vlib/v/checker/tests/if_mismatch_option_err.out
+++ b/vlib/v/checker/tests/if_mismatch_option_err.out
@@ -1,0 +1,12 @@
+vlib/v/checker/tests/if_mismatch_option_err.vv:8:2: warning: unused variable: `operation`
+    6 | fn main() {
+    7 |     operation_name := 'get_area'
+    8 |     operation := if o := Operation.from(operation_name) { o } else { ?Operation(none) }
+      |     ~~~~~~~~~
+    9 | }
+vlib/v/checker/tests/if_mismatch_option_err.vv:8:15: error: mismatched types `Operation` and `?Operation`
+    6 | fn main() {
+    7 |     operation_name := 'get_area'
+    8 |     operation := if o := Operation.from(operation_name) { o } else { ?Operation(none) }
+      |                  ~~
+    9 | }

--- a/vlib/v/checker/tests/if_mismatch_option_err.vv
+++ b/vlib/v/checker/tests/if_mismatch_option_err.vv
@@ -1,0 +1,9 @@
+enum Operation {
+	get_area
+	get_sector
+}
+
+fn main() {
+	operation_name := 'get_area'
+	operation := if o := Operation.from(operation_name) { o } else { ?Operation(none) }
+}

